### PR TITLE
Add e2e network policy tests.

### PR DESCRIPTION
**Pull request by ozdanborne**
**What this PR does / why we need it**:
Add set of e2e tests for Network Policy. This has succesfully run againast a deployment using Calico as the network policy provider.

Specifically, adds a new e2e test file (/test/e2e/network_policy.go) which tests TCP connectivity between pods with isolation.

See #25638 for PR that added NetworkPolicy resource.

This PR is a replacement for #27447 .

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35660)

### CODE REVIEW

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: A more descriptive name here would be nice - something like `Name: "allow-client-a"` maybe?

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: nit: extra space before `but`

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: Comment copy-paste error

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: Same here - better names are better :)

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: Strictly speaking, this test doesn't need isolation on Namespace B (since it's just testing policy in Namespace A).

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: One more: "test-policy"

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: extra space at beginning of comment.

s/do to/due to

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: comment doesn't match the test.

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: Instead of naming this after `i`, could you name it after the port it is listening on?

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: comment doesn't match the code is describes.

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: naming nit: `createServerPodAndService`

**Review by caseydavenport — CHANGES_REQUESTED** [review]
@djosborne Thanks for doing this!

A few nitty comments from me, but overall this looks like a nice set of basic e2es for NetworkPolicy.

**Comment by caseydavenport** [comment]
@djosborne You need to sign the linuxfoundation CLA.

File: test/e2e/network_policy.go
Comment (djosborne) [inline test/e2e/network_policy.go]: I don't see how that's true? This is the PodSelector test, which allows connections from pods with label `"pod-name": "client-a"`

Comment (caseydavenport) [inline test/e2e/network_policy.go]: D'oh, I somehow put a "not" in the comment when reading it.  Sorry...

**Comment by djosborne** [comment]
I signed it!

**Comment by djosborne** [comment]
Worth noting, I used the `etcd` container as it was the only one that had latest netcat installed, and doesn't suffer from [the dreaded search domain issue with alpine](https://github.com/gliderlabs/docker-alpine/issues/8).

Might be worth adding a busybox alternative / fix for e2e tests to use so I don't have to hijack etcd.

**Comment by caseydavenport** [comment]
@kubernetes/sig-network

**Comment by caseydavenport** [comment]
@k8s-bot ok to test

**Comment by caseydavenport** [comment]
@djosborne Looks like you've got some build failures.

```
W1028 13:10:07.907] test/e2e/network_policy.go:453: f.Client undefined (type *framework.Framework has no field or method Client)
```

**Comment by danwinship** [comment]
> Worth noting, I used the etcd container as it was the only one that had latest netcat installed, and doesn't suffer from the dreaded search domain issue with alpine.
> 
> Might be worth adding a busybox alternative / fix for e2e tests to use so I don't have to hijack etcd.

We also have LaunchWebserverPod() / CheckConnectivityToHost() in test/e2e/framework/util.go for this... (Well, it looks like they're currently unused, but this is what they were intended for; they originally came from the OpenShift network isolation tests.)

File: test/e2e/network_policy.go
Comment (danwinship) [inline test/e2e/network_policy.go]: "closed" sounds like it means "has had close() called on it". Something like "blocked" or "isolated" would be better. (And "unblocked"/"unisolated" rather than "opened" before)

File: test/e2e/network_policy.go
Comment (danwinship) [inline test/e2e/network_policy.go]: This is the only test case that passes multiple ports to `createServerPodAndService`, but nothing tests that `createServerPodAndService` actually works correctly in that case. (If there is a bug in `createServerPodAndService` such that it only opens the first port, this test would still pass.) And then if that happened, then the test would *always* pass, even if the NetworkPolicy implementation was broken, because it doesn't distinguish client-b failing because of isolation vs client-b failing because of connection refused.

Simple fix: flip the policy around so port2 is allowed and port1 is denied, and test that instead (and assume that `createServerPodAndService` isn't likely to be buggy in a way that would let this pass erroneously)
Better fix: test that both listeningPort1 and listeningPort2 work before applying the NetworkPolicy

File: test/e2e/network_policy.go
Comment (danwinship) [inline test/e2e/network_policy.go]: No, Casey's original comment was correct; line 347 says:

    // Create a pod with name 'client-a', which should be able to communicate with server.

and 348 says:

     By("Creating client-a in ns-a which should *not* be able to contact the server.")

(or maybe the comment got moved around after new commits were pushed...)

File: test/e2e/network_policy.go
Comment (danwinship) [inline test/e2e/network_policy.go]: and this comment is backwards too

**Comment by danwinship** [comment]
Additional test possibilities:
- "allow all" (eg, last example in https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/network-policy.md)
- multiple NetworkPolicies (eg, adding a second NetworkPolicy to a namespace allows additional connections, as opposed to being ignored, or replacing the first NetworkPolicy)
- adding a NetworkPolicy object to a namespace without the isolation annotation does not implicitly cause a DefaultDeny?

**Comment by djosborne** [comment]
Sorry for the delay, thanks for the review guys. Will revisit and clean up next week.

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: (in these tests) is redundant - we're already within the context of these tests :)

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: This "By" isn't great - it might not be a specific port that is opened.

Would be more clear if it was:

```
"Creating client pod %s that should successfully connect to %s", podName, service.Name)
```

Comment (caseydavenport) [inline test/e2e/network_policy.go]: Sae for testCannotConnect

File: test/e2e/pods.go
Comment (caseydavenport) [inline test/e2e/pods.go]: Undo this guy :)

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: s/stacked policy/policies

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: pods _cannot_ connect.

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: "only to _another_ port"

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: The name for this policy is wrong.

File: test/e2e/network_policy.go
Comment (caseydavenport) [inline test/e2e/network_policy.go]: comment is wrong.

**Comment by caseydavenport** [comment]
@djosborne a few more comments from me, but generally looking good!

**Comment by caseydavenport** [comment]
@djosborne looks like you're failing the verify target:

```
I1111 14:15:48.591] FAILED   hack/make-rules/../../hack/verify-bazel.sh	24s
I1111 14:15:48.591] Verifying hack/make-rules/../../hack/verify-boilerplate.sh
I1111 14:15:49.821] SUCCESS  hack/make-rules/../../hack/verify-boilerplate.sh	1s
I1111 14:15:49.821] Verifying hack/make-rules/../../hack/verify-cli-conventions.sh
I1111 14:16:10.837] +++ [1111 14:16:10] Generating bindata:
I1111 14:16:10.838]     /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
I1111 14:16:11.498] +++ [1111 14:16:11] Building the toolchain targets:
... skipping 516 lines ...
I1111 14:36:04.976] NetworkPolicy should enforce policy based on PodSelector
I1111 14:36:04.976] NetworkPolicy should enforce policy based on Ports
I1111 14:36:04.976] NetworkPolicy should support allow-all policy
I1111 14:36:04.976] NetworkPolicy should support setting DefaultDeny namespace policy
I1111 14:36:04.976] NetworkPolicy shouldn't enforce policy when isolation is off
I1111 14:36:04.977] 
I1111 14:36:04.977] ERROR: the test list has changed
I1111 14:36:04.981] Run ./hack/update_owners.py to fix it
I1111 14:36:04.983] SUCCESS  hack/make-rules/../../hack/verify-test-owners.sh	1s
I1111 14:36:04.984] Verifying hack/make-rules/../../hack/verify-flags-underscore.py
I1111 14:36:56.368] SUCCESS  hack/make-rules/../../hack/verify-flags-underscore.py	52s
I1111 14:36:56.369] Makefile:99: recipe for target 'verify' failed
W1111 14:36:56.369] make: *** [verify] Error 1
I1111 14:36:58.986] process 6872 exited with code 2
E1111 14:36:58.987] FAIL: pull-kubernetes-verify
```

**Comment by caseydavenport** [comment]
Also got some build failures:

```
W1111 14:17:16.932] test/e2e/network_policy.go:476: cannot use podName (type string) as type func() in argument to ginkgo.By
W1111 14:17:16.932] test/e2e/network_policy.go:476: cannot use service.ObjectMeta.Name (type string) as type func() in argument to ginkgo.By
W1111 14:17:16.932] test/e2e/network_policy.go:495: cannot use podName (type string) as type func() in argument to ginkgo.By
W1111 14:17:16.932] test/e2e/network_policy.go:495: cannot use service.ObjectMeta.Name (type string) as type func() in argument to ginkgo.B
```

**Comment by djosborne** [comment]
Can someone help explain why `hack/update_owners.py` needed to add me as an owner to [so many files](https://github.com/kubernetes/kubernetes/pull/35660/files#diff-060ac06d07884eb1f909da2e3e259fe4) to pass verification? And do I need to do anything to address those kops failures?

**Comment by caseydavenport** [comment]
@djosborne needs a rebase.  Looks like there are conflicts in the test_owners.csv.

My guess is that it added you to those tests since they were new and didn't have an owner.  Hopefully once the rebase is done that will be cleaned up!

**Comment by caseydavenport** [comment]
@k8s-bot node e2e test this

**Comment by caseydavenport** [comment]
@djosborne looks like you've got some type mismatches causing this to fail:

```
W1125 16:03:41.855] test/e2e/network_policy.go:570: cannot use pod (type *"k8s.io/kubernetes/pkg/api/v1".Pod) as type *"k8s.io/kubernetes/pkg/api".Pod in return argument
W1125 16:03:41.855] test/e2e/network_policy.go:570: cannot use svc (type *"k8s.io/kubernetes/pkg/api/v1".Service) as type *"k8s.io/kubernetes/pkg/api".Service in return argument
W1125 16:03:41.855] test/e2e/network_policy.go:597: cannot use "k8s.io/kubernetes/pkg/api".Pod literal (type *"k8s.io/kubernetes/pkg/api".Pod) as type *"k8s.io/kubernetes/pkg/api/v1".Pod in argument to f.ClientSet.Core().Pods(namespace.ObjectMeta.Name).Create
W1125 16:03:41.855] test/e2e/network_policy.go:601: cannot use pod (type *"k8s.io/kubernetes/pkg/api/v1".Pod) as type *"k8s.io/kubernetes/pkg/api".Pod in return argument
W1125 16:03:41.856] test/e2e/network_policy.go:624: cannot use namespace (type *"k8s.io/kubernetes/pkg/api".Namespace) as type *"k8s.io/kubernetes/pkg/api/v1".Namespace in argument to f.ClientSet.Core().Namespaces().Update
```

**Comment by caseydavenport** [comment]
@djosborne this LGTM.  Could you squash the commits?

**Comment by caseydavenport** [comment]
@thockin how can I get a LGTM label on this bad boy?

Also, looks like the CLA bot may not have picked up after the "I signed it"?

**Comment by djosborne** [comment]
I signed it!

**Comment by caseydavenport** [comment]
@danwinship any more thoughts?

Otherwise, @thockin can we get a LGTM?

**Comment by danwinship** [comment]
> We also have LaunchWebserverPod() / CheckConnectivityToHost() in test/e2e/framework/util.go for this...

Hm... so did you not notice this comment or was there some reason you didn't want to use those?

**Comment by danwinship** [comment]
> @danwinship any more thoughts?

Other than the question about using the pre-existing client/server utility methods, nope. If we have more test ideas later we can just add them later.

**Comment by djosborne** [comment]
@danwinship The LaunchWebserverPod() / CheckConnectivityToHost() don't support everything included in the tests I've written. Specifically, they don't generate a service in front of each Webserver pod, nor do they assign labels to pods (which the tests use to test pod-label-based policy).

The tests I've submitted work as is, I'd much prefer to merge now.

**Comment by caseydavenport** [comment]
@k8s-bot node e2e test this

**Comment by djosborne** [comment]
hey @deads2k I've finally done it - all green. Ready to merge?

cc: @caseydavenport

**Comment by deads2k** [comment]
> hey @deads2k I've finally done it - all green. Ready to merge?
> 
> cc: @caseydavenport

My name got caught during the initial approver storm. I have approved, so when @caseydavenport lgtm's, you'll enter the queue.  We're about to fork for 1.6, so I'm inclined to leave this in 1.7.

**Comment by caseydavenport** [comment]
@thockin can we get a lgtm label on this guy so the merge bot will let it through?

**Comment by caseydavenport** [comment]
@k8s-bot ok to test

**Comment by caseydavenport** [comment]
@djosborne looks like the rebase broke things:

```
I0410 16:49:18.147] k8s.io/kubernetes/test/e2e/network_policy.go:22: can't find import: "k8s.io/kubernetes/pkg/apis/meta/v1"
```

**Comment by deads2k** [comment]
> can't find import: "k8s.io/kubernetes/pkg/apis/meta/v1"

`metav1 k8s.io/apimachinery/pkg/apis/meta/v1`

**Comment by fejta** [comment]
FYI @caseydavenport @djosborne @deads2k this merge broke the submit queue! Looks like a bad rebase of test_owners.csv

https://github.com/kubernetes/test-infra/issues/2630

**Pull request merged**
